### PR TITLE
[Bugfix][KV Cache] Emit BlockRemoved only after the last physical copy is evicted

### PR DESCRIPTION
**Pull request by skajre**
## Purpose

A prefix-cache block hash can map to multiple physical KV cache blocks. For example, when an identical block-aligned prompt is served repeatedly, vLLM may reuse earlier prefix blocks but recompute the final prompt block to obtain next-token logits. The recomputed blocks have the same hash while retaining different physical block IDs.

Previously, evicting any one of these physical copies emitted a `BlockRemoved` event. An external KV event consumer could therefore invalidate the hash even though another physical copy was still resident and available for prefix-cache lookup.

For example, consider two cached requests:

```text
Request A: H1 -> H2 -> H3
Request B: H1 -> H2 -> H3 -> H4 -> H5 -> H6 -> H7
```

Suppose `H3` is backed by two physical blocks, `P3-A` and `P3-B`. Evicting `P3-A` previously emitted `BlockRemoved(H3)` even though `P3-B` remained resident. The external cache view could then appear as:

```text
H1, H2, H4, H5, H6, H7
```

Because block hashes form a prefix chain, a new request matching `H1 -> ... -> H7` would be scored as a two-block cache hit rather than the seven-block hit actually available in the engine. A prefix-aware router could consequently choose the wrong engine, or trigger unnecessary recomputation or remote loading of `H3` through `H7`.

This PR makes `BlockRemoved` a logical cache-key disappearance event:

- Remove the requested physical block from `BlockHashToBlockMap` and determine how many copies remain under the same `BlockHashWithGroupId` key.
- Suppress `BlockRemoved` while at least one same-hash physical copy remains.
- Emit the existing `BlockRemoved` event only after the last physical copy in that KV cache group disappears.
- Keep copy accounting internal to `BlockPool`; the event schema and ZMQ wire format remain unchanged.
- Preserve group isolation. Copies are scoped by `(block_hash, group_idx)`.

The internal removal result remains available to hash relocation paths such as `move_block_hashes`, so all hashes formerly owned by the source physical block are re-pointed correctly.

## Test Plan

```bash
pytest -q \
  tests/v1/core/test_prefix_caching.py \
  tests/v1/core/prefix_cache/test_partial_prefix_cache_primitives.py \
  tests/distributed/test_kv_cache_events.py \
  tests/distributed/test_events.py \
  tests/v1/kv_connector/unit/offloading_connector/test_events.py

ruff check \
  vllm/v1/core/block_pool.py \
  tests/v1/core/test_prefix_caching.py

ruff format --check \
  vllm/v1/core/block_pool.py \
  tests/v1/core/test_prefix_caching.py

git diff --check
```

End-to-end test:

1. Start a single-GPU Qwen3-8B full-attention server with prefix caching, block size 16, and an 8-block GPU cache.
2. Send three identical sequential 32-token completion requests without `prompt_logprobs`, producing three physical copies of the final block hash.
3. Send twelve unique 32-token requests to force eviction.
4. Consume events using the existing `BlockRemoved` schema.

## Test Result

```text
156 passed, 14 warnings in 26.84s
All checks passed
2 files already formatted
git diff --check: passed
```

End-to-end result:

```text
target_blockstored_before_pressure=3
target_blockremoved_before_pressure=0
target_blockremoved_after_pressure=1
wire_remaining_copy_counts_occurrences=0
subscriber_decode_errors=0
```

Observed target event stream:

```text
BlockStored(target_hash)  x 3
BlockRemoved(target_hash) x 1
```

No documentation update is required because this change preserves the public event schema and strengthens the existing `BlockRemoved` semantics.

### CODE REVIEW

**Comment by ai-jz** [comment]
Hi @skajre — thanks for working on this. Emitting `BlockRemoved` only after the last physical copy is evicted looks important for keeping KV-event consumers accurate.

Are you planning to continue this PR? I’d be happy to help with testing or follow-up changes if useful.

**Comment by skajre** [comment]
> Hi @skajre — thanks for working on this. Emitting `BlockRemoved` only after the last physical copy is evicted looks important for keeping KV-event consumers accurate.
> 
> Are you planning to continue this PR? I’d be happy to help with testing or follow-up changes if useful.

Thanks for reaching out! Yes, I plan to continue working on this PR and move the feature forward. I’ve already run the relevant tests around the BlockRemoved behavior, and the results look good so far. I’d be very happy to collaborate—additional testing or help with follow-up changes would be greatly appreciated. I’ll keep the PR updated as I make progress.

**Comment by skajre** [comment]
> Hi @njhill @ivanium, could you please take a look at this PR when you have time?
> It fixes premature BlockRemoved events when multiple physical KV blocks share the same hash, while preserving the existing event schema. The relevant unit and end-to-end tests are included and passing.
> I’d appreciate your feedback on the event semantics and the BlockPool implementation.

**Comment by ai-jz** [comment]
Thanks @skajre! We tested this together with #44488 in our SMG routing experiments; the [results and setup are in SMG #2497](https://github.com/smg-project/smg/pull/2497#issuecomment-5687971342). We haven't isolated the two fixes, but haven't found an additional issue with this PR in our testing so far.

Could you rebase onto the latest main to clear the conflicts?

@njhill @ivanium, could you take a look when you have a chance and help move this toward merge?

**Comment by skajre** [comment]
> Thanks @skajre! We tested this together with #44488 in our SMG routing experiments; the [results and setup are in SMG #2497](https://github.com/smg-project/smg/pull/2497#issuecomment-5687971342). We haven't isolated the two fixes, but haven't found an additional issue with this PR in our testing so far.
> 
> Could you rebase onto the latest main to clear the conflicts?
> 
> @njhill @ivanium, could you take a look when you have a chance and help move this toward merge?

Thanks for testing this in the SMG routing experiments and sharing the results!

I’ve rebased the PR onto the latest main, resolved the conflicts, and force-pushed the updated branch. The relevant test suite passes with 225 tests, and the PR is now mergeable.

The PR should be ready for another review. Thanks again for helping validate and move this forward!

File: vllm/v1/core/block_pool.py [hunk 1]
```diff
@@ -584,31 +592,39 @@ def _get_partial_block_parent_hash_and_start(
...
+            if removed_block is not None:
+                removed_hashes.append((block_hash, remaining_count))
         block.reset_hash()
         return removed_hashes
 
     def _emit_block_removed_events(
         self,
-        block_hashes: list[BlockHashWithGroupId],
+        block_hashes: list[tuple[BlockHashWithGroupId, int]],
     ) -> None:
```
Comment (ivanium) [inline vllm/v1/core/block_pool.py:618]: After re-reading the code, I am thinking we can potentially fold this func into `_remove_cached_block_hashes()` and simplify the code

File: vllm/v1/core/block_pool.py [hunk 2]
```diff
@@ -653,7 +669,7 @@ def move_block_hashes(
         assert dst_block.block_hash is None
         assert dst_block.block_id not in self.cached_block_hashes_by_block
         num_tokens = src_block.block_hash_num_tokens
-        for block_hash in self._remove_cached_block_hashes(src_block):
+        for block_hash, _ in self._remove_cached_block_hashes(src_block):
```
Comment (ivanium) [inline vllm/v1/core/block_pool.py:672]: if we fold _emit_block_removed_events() into _remove_cached_block_hashes(), we need to add a flag to _remove_cached_block_hashes() to skip emitting event here.

**Review by ivanium — COMMENTED** [review]
Thanks for the fix. Left comments on potential simplification.

File: vllm/v1/core/block_pool.py (hunk 1 above)
Comment (skajre) [reply vllm/v1/core/block_pool.py]: Done, thanks! I folded `_emit_block_removed_events()` into `_remove_cached_block_hashes()`, which now emits `BlockRemoved` only when the last same-hash copy is removed.

File: vllm/v1/core/block_pool.py (hunk 2 above)
Comment (skajre) [reply vllm/v1/core/block_pool.py]: Done. I added an `emit_events` flag and call `_remove_cached_block_hashes(..., emit_events=False)` from `move_block_hashes()`. I also added a regression test to verify that moving hashes does not emit a `BlockRemoved` event.
